### ISSUE TITLE
Add cloudprovider support

### ISSUE THREAD

**Issue by galal-hussein** [rancher/rke2#142]
RKE2 should support in-tree cloud providers by enabling flags to support both names and cloud config file and pass them to:

- kubelet
- kube-controller-manager
- kube-api

**Comment by rancher-max** [comment]
**A couple of things have come out of testing so far:**

1. **AWS in-tree provider works ONLY when setup properly.** What this means is if I supply the following command: `INSTALL_RKE2_VERSION=v1.18.4-beta8+rke2 INSTALL_RKE2_EXEC="--token secret --cloud-provider-name aws --write-kubeconfig-mode=644 --node-name=$(hostname -f)" ./install.sh` it works IF AND ONLY IF the EC2 Instance has a valid IAM role attached for cloudprovider and a tag for `kubernetes.io/cluster/<clusterId> = <owned | shared>`. If these are not in place, rke2 does not finish installing and I receive the following error on repeat in the logs:
```

Aug 26 15:29:45 ip-172-31-50-136 rke2[20304]: time="2020-08-26T15:29:45Z" level=error msg="Unable to watch for tunnel endpoints: Get \"https://127.0.0.1:6443/api/v1/namespaces/default/endpoints?fieldSelector=metadata.name%3Dkubernetes&resourceVersion=0&watch=true\": dial tcp 127.0.0.1:6443: connect: connection refused"
Aug 26 15:29:45 ip-172-31-50-136 rke2[20304]: {"level":"warn","ts":"2020-08-26T15:29:45.701Z","caller":"clientv3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"passthrough:///https://127.0.0.1:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = latest balancer error: connection error: desc = \"transport: Error while dialing dial tcp 127.0.0.1:2379: connect: connection refused\""}
Aug 26 15:29:45 ip-172-31-50-136 rke2[20304]: time="2020-08-26T15:29:45Z" level=info msg="Failed to test data store connection: context deadline exceeded"
Aug 26 15:29:48 ip-172-31-50-136 rke2[20304]: time="2020-08-26T15:29:48Z" level=error msg="Kubelet exited: exit status 255"
Aug 26 15:29:50 ip-172-31-50-136 rke2[20304]: time="2020-08-26T15:29:50Z" level=error msg="Unable to watch for tunnel endpoints: Get \"https://127.0.0.1:6443/api/v1/namespaces/default/endpoints?fieldSelector=metadata.name%3Dkubernetes&resourceVersion=0&watch=true\": dial tcp 127.0.0.1:6443: connect: connection refused"
Aug 26 15:29:54 ip-172-31-50-136 rke2[20304]: time="2020-08-26T15:29:54Z" level=error msg="Kubelet exited: exit status 255"
```

2. **Azure in-tree provider receives the same error as mentioned above**. I believe this is probably for one of two reasons: either I have set it up incorrectly, or there is a relation to a rancher issue: https://github.com/rancher/rancher/issues/28049 due to the way my instances are configured. When installing this, I use the following command: `INSTALL_RKE2_VERSION=v1.18.4-beta8+rke2 INSTALL_RKE2_EXEC="--token secret --cloud-provider-name azure --cloud-provider-config ./config.conf --write-kubeconfig-mode=644 --node-name=max-simple-vm-name" ./install.sh` where config.conf contains:
```
{
    "tenantId": "<app tenant id>",
    "aadClientId": "<app client id",
    "aadClientSecret": "<app client secret>",
    "subscriptionId": "<either subscription id noted on my instance OR subscription id from the app>"
}
```
I'm not sure how to grab the etcd container logs on either of these, but checking the containerd logs themselves (`cat /var/lib/rancher/rke2/agent/containerd/containerd.log`) does appear that it attempted to start etcd. 

3. **AWS out-of-tree provider generally works**. I followed the same steps as mentioned in https://github.com/rancher/k3s/issues/1807#issuecomment-675081920 (but replaced k3s related args with rke2. Differences being the install command used was: `INSTALL_RKE2_VERSION=v1.18.4-beta8+rke2 INSTALL_RKE2_EXEC="--token secret --kubelet-arg="cloud-provider=external" --kubelet-arg="provider-id=aws:///$provider_id" --write-kubeconfig-mode=644 --node-name=$(hostname -f)" ./install.sh` and the helmchart yaml was added to the same place but with "k3s" replaced by "rke2"). This led to the same results noted in that comment. When I did not have the IAM role and tag attached to the instance, the aws-cloud-controller-manager pod was in CrashLoopBackoff with logs:
```
I0826 16:08:18.194977       1 serving.go:313] Generated self-signed cert in-memory
W0826 16:08:19.455313       1 client_config.go:543] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
I0826 16:08:19.461942       1 controllermanager.go:120] Version: v0.0.0-master+$Format:%h$
W0826 16:08:19.461972       1 plugins.go:115] WARNING: aws built-in cloud provider is now deprecated. The AWS provider is deprecated and will be removed in a future release
I0826 16:08:19.463846       1 aws.go:1214] Building AWS cloudprovider
I0826 16:08:19.463911       1 aws.go:1180] Zone not specified in configuration file; querying AWS metadata service
F0826 16:08:19.613062       1 controllermanager.go:124] Cloud provider could not be initialized: could not init cloud provider "aws": error finding instance i-011a2866be14112c7: "error listing AWS instances: \"NoCredentialProviders: no valid providers in chain. Deprecated.\\n\\tFor verbose messaging see aws.Config.CredentialsChainVerboseErrors\""
```
But after adding the role and tag, this was resolved and the ingress-nginx and metrics-server helm install and rke2-coredns pods (previously stuck in Pending) were Completed and Running successfully.

**Comment by brandond** [comment]
1. I think we need to figure out why etcd doesn't come up in scenarios 1 and 2. That is unexpected and definitely a problem.
1. For scenario 3, I'm curious about is error: `plugins.go:115] WARNING: aws built-in cloud provider is now deprecated.` - this seems to indicate that the in-tree cloud provider is still being used or is at least trying to start up.

**Comment by davidnuzik** [comment]
@rancher-max What else is needed to get this moving forward? Could you summarize current status on this issue?

**Comment by rancher-max** [comment]
Essentially there is an issue with Azure that is the same as mentioned in https://github.com/rancher/rancher/issues/28049. I believe otherwise it is working, but have not been able to fully validate that yet. In summary:
**Validated:**
- In-tree WITHOUT config file (aws)
- Out-of-tree
**Pending:**
- In-tree WITH config file (azure)

If the issue gets fixed, then I can validate azure. I believe I can validate at least the overall functionality using one subscription before the multi-subscription issue gets fixed and will be working on that soon.

**Comment by davidnuzik** [comment]
@rancher-max if what issue gets fixed? https://github.com/rancher/rancher/issues/28049 appears fixed and closed.

**Comment by rancher-max** [comment]
I just created a new issue for it in this repo: https://github.com/rancher/rke2/issues/318

**Comment by davidnuzik** [comment]
If good with AWS can close this out.

**Comment by rancher-max** [comment]
Closing per above comment -- AWS looks good and there is a separate issue for Azure in-tree using config file.
